src/main/java/io/nats/client/impl/Headers.java

```diff
@@ -0,0 +1,346 @@
+// Copyright 2015-2018 The NATS Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package io.nats.client.impl;
+
+import java.util.*;
+
+import static java.nio.charset.StandardCharsets.US_ASCII;
+
+
+public class Headers {
+	private static String COLON = ":";
+	private static String COLON_SPACE = ": ";
+	private static String CRLF = "\r\n";
+	private static String HEADER_PREFIX = "NATS/1.0\r\n";
+
+	private static final String KEY_CANNOT_BE_EMPTY_OR_NULL = "Header key cannot be null.";
+	private static final String KEY_INVALID_CHARACTER = "Header key has invalid character: ";
+	private static final String VALUE_INVALID_CHARACTERS = "Header value has invalid character: ";
+
+	private static final boolean normalizeKey = false;
+
+	private Map<String, List<String>> headerMap = new HashMap<>();
```

Comment:
This can probably be changed to:
```
private Map<ByteBuffer List<ByteBuffer>> headerMap = new HashMap<>();
```
From what I can tell the headers are ASCII which you can operate on without any string conversions generally.

Reply:
Let's stick with String for now to be consistent with other languages (e.g. .NET uses string), and then take a pass at optimization afterword.  We might end up supporting UTF-8 in headers, TBD...

Reply:
I'm mostly just referring to the internal data structures since Java doesn't have support for native UTF-8 or ASCII strings, public API's can still accept/return strings and convert to/from an internal ByteBuffer format.

Reply:
A couple things here. If we use ByteBuffer as a key, I would have to write a hashcode and equals, because they don't exist (which means they default to the Object's). As you mentioned, developers are going to want to work in Strings and if we provide a string api with ByteBuffers underneath, we would have to convert every time we put it into the map.
@jameshilliard Can you look at the 2 things, first the Headers class, specifically the method getSerialized. So when the publish code asks for the headers, it asks for them serialized, which I give directly in a byte[], which is generated using ByteBuffers under the covers in ByteArrayBuilder. The second thing, could you look at ByteArrayBuilder, I'd like some comment on how I'm calculating expanding the buffer.

Reply:
> If we use ByteBuffer as a key, I would have to write a hashcode and equals, because they don't exist (which means they default to the Object's). 

They both exist, see [equals](https://docs.oracle.com/javase/8/docs/api/java/nio/ByteBuffer.html#equals-java.lang.Object-) and [hashCode](https://docs.oracle.com/javase/8/docs/api/java/nio/ByteBuffer.html#hashCode--). Note that #356 makes use of these features for the hashmaps and buffer matching.

> As you mentioned, developers are going to want to work in Strings and if we provide a string api with ByteBuffers underneath, we would have to convert every time we put it into the map.

Sure, but they eventually will have to be serialized to a bytes like type at least once at a minimum, IMO better to do that up front and use a normalized `ByteBuffer` format internally.

Reply:
> The second thing, could you look at ByteArrayBuilder, I'd like some comment on how I'm calculating expanding the buffer.

This class to me seems like a hack to add inefficient string conversion/resize operations to `ByteBuffer`'s, these operations doesn't seem necessary as long as you ensure each of the individual header `Strings` are already converted to `ByteBuffer`'s by the time they need to be serialized as you can then do a single up front allocation by checking `ByteBuffer` sizes using `remaining()`. Header serialization should probably be done in `NatsMessage`, maybe by passing a headers object to it, this way you don't have to do any separate buffer allocations for the serialized headers. See how I do the full message serialization with a single allocation [here](https://github.com/nats-io/nats.java/pull/356/files#diff-519f371d84d8e0274587b137b1f2fcfcaa9ab8afb0897e54bcc4418e0b2f7244R52) and then just create shared references to it for the field specific slice buffers

Reply:
Basically the way I've been refactoring this is by first ensuring that everything is already converted to a `ByteBuffer` by the time you need to calculate a buffer size, as pre-calculating a buffer size based on a string is not efficient and difficult as Java does not support ASCII/UTF-8 encoded strings natively.

Reply:
Here's what I mean: https://github.com/nats-io/nats.java/pull/364/commits/f833b99d51166821afcd544d29a0f32a84d94ace

Reply:
Notice that I'm using the "`serialized`" `ByteBuffer` to back the `headerMap` as well to reduce allocations.

Reply:
Notice that I only do a single buffer allocation, the one in `getSerialized` which allocates the exact buffer size needed for header serialization up front, even this single allocation can probably be eliminated by writing directly to a `NatsMessage` buffer.

Reply:
Not sure how I missed the equals, but you are right it's there. Can you benchmark the two implementations?

Also regarding the ByteArrayBuffer, we'll have to agree to disagree. There are plenty of string conversions for the keys and values to use them map, so I'm not seeing a huge benefit there. I also think in practical use, there will be minimal resizing since the original allocation can be set and the default probably covers a ton of cases, for instance 32 or 64 bytes for a protocol message, and then maybe one resize to hold the entire data block. I'm sure there are optimization as well when we factor in usage, but that's outside of the scope of getting headers working.

The headers version with byte buffers is 100 lines longer (379 from 275), mostly for the serialization. The NatsMessage code where I use ByteArrayBuilder is much more compact / readable and saves a lot of thinking ahead to figure out the lengths ahead of time.

I feel like this code is more readable, more maintainable and fast enough - for a client. In fact my addition of using byte buffers was already to respect your position on using them. I suppose for the server, when every single nano counts I would more wholeheartedly support this level of effort (which is why it's written in go,) but again for the client?

Reply:
> Can you benchmark the two implementations?

When I did the initial refactoring of `NatsMessage` to use `ByteBuffer`'s avoiding unnecessary buffer allocations had a significant performance benefit.

> There are plenty of string conversions for the keys and values to use them map, so I'm not seeing a huge benefit there.

The benefit is that you should not be expanding the `ByteBuffer` as that requires expensive reallocations, string conversions should all be done prior to buffer allocation, rather than repeatedly reallocating and expanding the buffer.

> The NatsMessage code where I use ByteArrayBuilder is much more compact / readable and saves a lot of thinking ahead to figure out the lengths ahead of time.

I noticed a number of rather subtle buffer sizing issues with how you were serializing buffers, I'm trying to sidestep these issues entirely.

> I feel like this code is more readable

It's less verbose maybe but has subtle edge cases and performance issues.

> I suppose for the server, when every single nano counts I would more wholeheartedly support this level of effort (which is why it's written in go,) but again for the client?

I'm planning to use this for a protocol converting proxy gateway server in Java.

I went ahead and refactored out the usage of `ByteArrayBuilder` here already and fixed the failing tests: https://github.com/nats-io/nats.java/pull/364/commits/92e4c6c9515c3e6ee728fd40b2605aff093ccb61

Reply:
Here's some benchmarks of the `NatsMessage` protocol serialization codepath before and after removing `ByteArrayBuilder` usage:
Before:
```
### Total time to create 50,000,000 non-utf8 messages for sending was 9,724 ms
	194.491932 ns/op
	5,141,601.44 op/sec

### Total time to create 50,000,000 utf8 messages for sending was 10,006 ms
	200.134972 ns/op
	4,996,627.964 op/sec
```
After:
```
### Total time to create 50,000,000 non-utf8 messages for sending was 8,437 ms
	168.749313 ns/op
	5,925,950.062 op/sec

### Total time to create 50,000,000 utf8 messages for sending was 8,598 ms
	171.961319 ns/op
	5,815,261.283 op/sec
```

Reply:
> In fact my addition of using byte buffers was already to respect your position on using them.

You have to be careful with how you use `ByteBuffer`'s, I'm primarily using them to reduce buffer copy/allocation operations and improve compatibility with the buffer passing asynchronous socket API's by natively operating on `ByteBuffers` within serialization/de-serialization classes. The approach you've used in `ByteArrayBuilder` to encapsulate `ByteBuffer` operations goes counter to those goals as you're just converting them in the end to regular `byte[]` arrays and are also greatly increasing copying/allocations via resizing.

Reply:
Basically you want to look at each `ByteBuffer.allocate()` as a c style char buffer `malloc()`. The `ByteBuffer` class is just a thin abstraction over a c style char buffer and some pointers from my understanding.

Reply:
I'm not under allocating / resizing too often, I've already addressed that. So my takeaway from this is to reduce buffer copy. For instance when serializing the headers and other uses of ByteArrayBuilder, return or allow access to the ByteBuffer to allow the copy to come from the ByteBuffer's byte[] instead of the copy to come from the byte[] created from the ByteBuffer.
In my opinion, it's not just less verbose, it's significantly simpler. As in all software there are trade offs, and, again, my opinion is that this trade off is well worth it for less than 2 second loss over 50 million passes for the message. For our regular use case, at least for today when the priority is getting functional headers, this is more than acceptable, I cannot speak to your specific use case.

Reply:
> I'm not under allocating / resizing too often, I've already addressed that.

You shouldn't be resizing those buffers at all if you allocate the correct amount up front, which is straight forward to do.

> In my opinion, it's not just less verbose, it's significantly simpler.

Not disagreeing with that, but that's not the issue, the issue is that it creates a significant performance regression by not sizing correctly up front.

> As in all software there are trade offs, and, again, my opinion is that this trade off is well worth it for less than 2 second loss over 50 million passes for the message.

Well it all adds up, the more places `ByteArrayBuilder` gets used instead of up front correctly sized buffers the worse performance will get.

> For our regular use case, at least for today when the priority is getting functional headers, this is more than acceptable, I cannot speak to your specific use case.

I have headers with up front correctly sized buffers working in #364(which should apply cleanly on top of this PR) so not sure why it would make sense to use the slower version.

Reply:
> As in all software there are trade offs, and, again, my opinion is that this trade off is well worth it for less than 2 second loss over 50 million passes for the message.

So it seems it's actually a much larger difference than I even expected, in a number of the autobench benchmarks I'm showing using correctly upfront sized buffers as ~50% faster than the `ByteArrayBuilder` variant which is a larger difference than I'm seeing in just the message serialization benchmarks by themselves, not sure why that is but an overall 50% performance gain by correctly sizing the buffers up front seems worth some extra complexity to me.

Reply:
> In my opinion, it's not just less verbose, it's significantly simpler. 

Your latest changes made it much easier for me to adapt the internals of `Headers` to use `ByteBuffer`'s cleanly while still doing precise buffer allocations and zero-copy parsing of headers.